src/csharp/Microsoft.Spark.Worker/Command/SqlCommandExecutor.cs

```diff
@@ -118,28 +117,38 @@ internal class PicklingSqlCommandExecutor : SqlCommandExecutor
                             $"Invalid message length: {messageLength}");
                     }
 
-                    MaxLengthReadStream readStream = s_slicedReadStream ??
-                            (s_slicedReadStream = new MaxLengthReadStream());
-                    readStream.Reset(inputStream, messageLength);
-
-                    // Each row in inputRows is of type object[]. If a null is present in a row
-                    // then the corresponding index column of the row object[] will be set to null.
-                    // For example, (inputRows.Length == 2) and (inputRows[0][0] == null)
-                    //   +----+
-                    //   | age|
-                    //   +----+
-                    //   |null|
-                    //   |  11|
-                    //   +----+
-                    object[] inputRows = PythonSerDe.GetUnpickledObjects(readStream);
+                    byte[] buffer = ArrayPool<byte>.Shared.Rent(messageLength);
+                    object[] inputRows = null;
+
+                    try
+                    {
+                        if (!SerDe.TryReadBytes(inputStream, buffer, messageLength))
+                        {
+                            throw new IOException("premature end of input stream");
+                        }
+                        // Each row in inputRows is of type object[]. If a null is present in a row
+                        // then the corresponding index column of the row object[] will be set to null.
+                        // For example, (inputRows.Length == 2) and (inputRows[0][0] == null)
+                        //   +----+
+                        //   | age|
+                        //   +----+
+                        //   |null|
+                        //   |  11|
+                        //   +----+
+                        inputRows = PythonSerDe.GetUnpickledObjects(new ReadOnlyMemory<byte>(buffer, 0, messageLength));
+                    }
+                    finally
+                    {
+                        ArrayPool<byte>.Shared.Return(buffer);
+                    }
 
                     for (int i = 0; i < inputRows.Length; ++i)
                     {
                         // Split id is not used for SQL UDFs, so 0 is passed.
                         outputRows.Add(commandRunner.Run(0, inputRows[i]));
                     }
 
-                    WriteOutput(outputStream, outputRows);
+                    WriteOutput(outputStream, outputRows, messageLength);
```

Comment:
The input `messageLength` doesn't really have anything to do with the output message length. This feels like an odd association/heuristic.

Reply:
The output size should be <= input (I have verified that for Q1, Q8 and Q13). It's going to be allocated once and then reused many times. That was the best estimation I could come up with based on the available data.

Reply:
Sounds good to me.
Can we add a comment here with that reasoning?

Reply:
Sure, I did it right after you asked this question.

https://github.com/dotnet/spark/blob/674a83d1d146e2053d573fda15fdc2a4c2b2d885/src/csharp/Microsoft.Spark.Worker/Command/SqlCommandExecutor.cs#L136